src/common/range.rs

```diff
@@ -34,96 +42,93 @@ use std::ops::{Bound, RangeBounds};
+    /// 4.1 of RFC7233](https://tools.ietf.org/html/rfc7233#section-4.1).
+    Bytes(Vec<ByteRangeSpec>),
+    /// Range units are intended to be extensible. New range units ought to be
+    /// registered with IANA, as defined in [Section 5.1 of RFC 7233](2).
+    ///
+    /// `bytes` is the only [registered](1) unit at the moment, so all other
+    /// units will end up in this variant, `Unregistered`. For those other
+    /// units, no concrete format for the `other-ranges-set` is given in RFC
+    /// 7233, so additional parsing has to be done by the consumer of the
+    /// unregistered units.
+    ///
+    /// [1]: https://www.iana.org/assignments/http-parameters/http-parameters.xhtml#range-units
+    /// [2]: https://datatracker.ietf.org/doc/html/rfc7233#section-5.1
+    Unregistered {
+        /// The unit for other ranges. This has to be a `token`, as defined in
+        /// [section 3.2.6 of RFC 7230](1).
+        ///
+        /// [1]: https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6
+        unit: String,
+        /// Representation of the range set for the custom, unregistered unit.
+        /// This can be a string made up out of any printable ASCII character.
+        set: String,
+    },
 }
 
-fn parse_bound(s: &str) -> Option<Bound<u64>> {
-    if s.is_empty() {
-        return Some(Bound::Unbounded);
-    }
-
-    s.parse().ok().map(Bound::Included)
+/// A single byte range, which can either be a range or a suffix.
+#[derive(Clone, Debug, PartialEq)]
+pub enum ByteRangeSpec {
+    /// For a range, indices of the first and last byte are given.
+    FromTo(u64, u64),
+    /// Remainder of the resource, starting from the byte index.
+    AllFrom(u64),
+    /// For a suffix, only the length of the suffix is given, and the server is prompted to return
+    /// the last bytes only.
+    Last(NonZeroU64),
 }
 
-impl ::Header for Range {
-    fn name() -> &'static ::HeaderName {
-        &::http::header::RANGE
-    }
-
-    fn decode<'i, I: Iterator<Item = &'i ::HeaderValue>>(values: &mut I) -> Result<Self, ::Error> {
-        values
-            .next()
-            .and_then(|val| {
-                if val.to_str().ok()?.starts_with("bytes=") {
-                    Some(Range(val.clone()))
-                } else {
-                    None
-                }
-            })
-            .ok_or_else(::Error::invalid)
-    }
-
-    fn encode<E: Extend<::HeaderValue>>(&self, values: &mut E) {
-        values.extend(::std::iter::once(self.0.clone()));
+impl Range {
+    /// Convenience method to convert into [RangeBounds], if this is a set of byte ranges.
+    /// If it isn't, then this will return `None`.
+    pub fn to_range_bounds(&self, len: u64) -> Option<Vec<impl RangeBounds<u64>>> {
+        match self {
+            Self::Bytes(ranges) => Some(
+                ranges
+                    .iter()
+                    .filter_map(|spec| spec.to_range_bounds(len))
```

Comment:
Ah yes, that method name should probably be a bit more verbose, and looking at it again, it's not *quite* correct either, because it still returns some if the Vec is empty. Will fix :)

Reply:
mhm, thought about this a bit more (after I ran into trouble with returning None on empty vec): The option is to check whether it's a range header that makes sense to be represented by range bounds, so maybe `Some` together with an empty Vec *does* make sense. Again, more thoughts on this would be appreciated.

Reply:
As far as I can tell from RFC 7233, there's no explicit behaviour specified for unrecognized range units. With that in mind, it's probably best to keep this method as simple as possible and let the caller decide how they want to handle it. think you're right about returning `Some` of empty `Vec`.

Reply:
I've changed the return type of both `Range::to_satisfiable_range_bounds` and `ByteRangeSpec::to_satisfiable_range_bounds` from `Option` to `Result`. I'm using `::Error` as the error type here although I would prefer a custom error type here indicating that the error happened during conversion to end inclusive range bounds. As I haven't seen any custom error types apart from `::Error` and I'm not sure what the right convention would be for custom error types, I'm just using `::Error` here.